### DISCUSSION TITLE
Suggestion: Plugin system

### DISCUSSION THREAD

**Discussion by javierav** [basecamp/kamal#258] [General]
Are you considering the possibility of implementing a plugin system using rubygems? This would allow extending the functionality of mrsk just by installing additional gems.

For example the hypothetical `mrsk-mysql` gem could add the mrsk mysql command and this would implement additional subcommands to manage mysql, such as creating users or databases for projects.

I'm thinking something similar to https://github.com/dokku/dokku-mysql

**Comment by scart88** [comment]
Or trigger manual or scheduled backups, restores from those backups...

**Comment by dhh** [comment]
Thor doesn't lend itself easily to this, unlike Rake. But I would like to have some more callback hooks, ala what we have for the broadcast command. This should all be driven by real, specific needs, though. I think the MySQL example is a fine one to start with.

**Comment by dhh** [comment]
We now have hooks, which provide similar service. Can extend as real use cases are proposed.

**Comment by koenhandekyn** [comment]
for another use case for the plugin system - see  https://github.com/basecamp/kamal/discussions/1240#discussioncomment-11337294 

i fail to see however how hooks fit in to the mentioned plug in system for extending commands/capabilities, e.g. for mysql, postgres, ...
